gdpr/impl.go

```diff
@@ -74,38 +77,99 @@ func (p *permissionsImpl) allowSync(ctx context.Context, vendorID uint16, consen
 	if consent == "" {
-		return p.cfg.UsersyncIfAmbiguous, nil
+		return p.cfg.UsersyncIfAmbiguous, p.cfg.UsersyncIfAmbiguous, nil
 	}
 
 	parsedConsent, vendor, err := p.parseVendor(ctx, vendorID, consent)
 	if err != nil {
-		return false, err
+		return false, false, err
 	}
 
 	if vendor == nil {
-		return false, nil
+		return false, false, nil
 	}
 
 	if parsedConsent.Version() == 2 {
-		// Need to add the location special purpose once the library supports it.
+		if p.cfg.TCF2.Enabled {
+			return p.allowPITCF2(parsedConsent, vendor, vendorID)
+		}
 		if (vendor.Purpose(consentconstants.InfoStorageAccess) || vendor.LegitimateInterest(consentconstants.InfoStorageAccess)) && parsedConsent.PurposeAllowed(consentconstants.InfoStorageAccess) && (vendor.Purpose(consentconstants.PersonalizationProfile) || vendor.LegitimateInterest(consentconstants.PersonalizationProfile)) && parsedConsent.PurposeAllowed(consentconstants.PersonalizationProfile) && parsedConsent.VendorConsent(vendorID) {
-			return true, nil
+			return true, true, nil
 		}
 	} else {
 		if (vendor.Purpose(tcf1constants.InfoStorageAccess) || vendor.LegitimateInterest(tcf1constants.InfoStorageAccess)) && parsedConsent.PurposeAllowed(tcf1constants.InfoStorageAccess) && (vendor.Purpose(tcf1constants.AdSelectionDeliveryReporting) || vendor.LegitimateInterest(tcf1constants.AdSelectionDeliveryReporting)) && parsedConsent.PurposeAllowed(tcf1constants.AdSelectionDeliveryReporting) && parsedConsent.VendorConsent(vendorID) {
-			return true, nil
+			return true, true, nil
 		}
 	}
-	return false, nil
+	return false, false, nil
+}
+
+func (p *permissionsImpl) allowPITCF2(parsedConsent api.VendorConsents, vendor api.Vendor, vendorID uint16) (allowPI bool, allowGeo bool, err error) {
+	consent, ok := parsedConsent.(tcf2.ConsentMetadata)
+	err = nil
+	allowPI = false
+	allowGeo = false
+	if !ok {
+		err = fmt.Errorf("Unable to access TCF2 parsed consent")
+		return
+	}
+	if p.cfg.TCF2.SpecialPurpose1 {
+		allowGeo = consent.SpecialFeatureOptIn(1) && vendor.SpecialPurpose(1)
+	} else {
+		allowGeo = true
+	}
+	// Set to true so any purpose check can flip it to false
+	allowPI = true
+	if p.cfg.TCF2.Purpose1 {
+		allowPI = allowPI && p.checkPurpose(consent, vendor, vendorID, consentconstants.InfoStorageAccess)
+	}
+	if p.cfg.TCF2.Purpose2 {
+		allowPI = allowPI && p.checkPurpose(consent, vendor, vendorID, consentconstants.BasicAdserving)
+	}
+	if p.cfg.TCF2.Purpose4 {
+		allowPI = allowPI && p.checkPurpose(consent, vendor, vendorID, consentconstants.PersonalizationSelection)
+	}
+	if p.cfg.TCF2.Purpose7 {
+		allowPI = allowPI && p.checkPurpose(consent, vendor, vendorID, consentconstants.AdPerformance)
+	}
+	return
+}
+
+func (p *permissionsImpl) checkPurpose(consent tcf2.ConsentMetadata, vendor api.Vendor, vendorID uint16, purpose tcf1constants.Purpose) bool {
+	if purpose == 1 && p.cfg.TCF2.PurposeOneTreatment.Enabled && consent.PurposeOneTreatment() {
```

Comment:
Can you please explain what the trailing `consent.PurposeOneTreatment()` adds to the logic?

Reply:
First comparison is to ensure we are considering purpose 1. The second comparison is to see if the PBS Host as configured to care about the PurposeOneTreatment flag. The last, `consent.PurposeOneTreatment()` checks to see if the flag is set in the consent string. It is last because it is the most expensive check.